packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js

```diff
@@ -1381,50 +1392,97 @@ function collectRecommendations({
   };
 }
 
-// Finds functions declared as dependencies
+// If the node will result in constructing a referentially unique value, return
+// its human readable type name, else return null.
+function getConstructionExpresionType(node) {
+  switch (node.type) {
+    case 'ObjectExpression':
+      return 'object';
+    case 'ArrayExpression':
+      return 'array';
+    case 'ArrowFunctionExpression':
+    case 'FunctionExpression':
+      return 'function';
+    case 'ClassExpression':
+      return 'class';
+    case 'ConditionalExpression':
+      if (
+        getConstructionExpresionType(node.consequent) != null ||
+        getConstructionExpresionType(node.alternate) != null
+      ) {
+        return 'conditional';
+      }
+      return null;
+    case 'LogicalExpression':
+      if (
+        getConstructionExpresionType(node.left) != null ||
+        getConstructionExpresionType(node.right) != null
+      ) {
+        return 'logical expression';
```

Comment:
For even better developer experience, we could try to blame the actual sub-expression that is a construction.

For example:

```
const foo = bar ?? [];
                   ^^ This is actually to blame
```

However, it would add some additional complexity.
```